### ISSUE TITLE
Flatten array values into single (n x 1) column 

### ISSUE THREAD

**Issue by gjvanvuuren** [Cinchoo/ChoETL#180]
Hi,

I'm having trouble doing a pivot of a flattened array into a column. Using one of your examples:

```
string json = @"
{
    ""user"": {
        ""name"": ""asdf"",
        ""teamname"": ""b"",
        ""email"": ""c"",
        ""players"": [""1"", ""2""]
    }
}";

var conf = new ChoJSONRecordConfiguration
{
	FlattenNode = true
};

using (var r = new ChoJSONReader(JObject.Parse(json), conf)
	.WithJSONPath("$.user")
)
{
	var dt = r.AsDataTable();
	Console.WriteLine(dt);
}
```

I'm trying to get the json into the following tabular form:

```
name, teamname, email, player
asdf, b, c, 1
asdf, b, c, 2
```

This should be simple, but I can't figure it out. Using the `valueConverter:` argument, I've seen how join the array values into a single row, but not how to do a cross join to get each value into a new row.

I'd appreciate your support.

Thanks

**Comment by Cinchoo** [comment]
Pls take v[1.2.1.43-beta6](https://www.nuget.org/packages/ChoETL.JSON.NETStandard/1.2.1.43-beta6), give it a try.

**Comment by gjvanvuuren** [comment]
Thanks, that works if there isn't another key-value on the same level as `user`, but if I change the json to:

```
string json = @"
{
   ""mykey"": 1234,
    ""user"": {
        ""name"": ""asdf"",
        ""teamname"": ""b"",
        ""email"": ""c"",
        ""players"": [""1"", ""2""]
    }
}";

```
then as soon as I remove `.WithJSONPath` in order to resolve `mykey`, the `players` array reverts back to the default multicolumn structure.

```
string json = @"
{
  ""mykey"": 1234,
  ""user"": {
      ""name"": ""asdf"",
      ""teamname"": ""b"",
      ""email"": ""c"",
      ""players"": [""1"", ""2""]
  }
}";

var conf = new ChoJSONRecordConfiguration
{
	FlattenNode = true
};

using (var r = ChoJSONReader.LoadText(json, conf)
)
{
	var dt = r.AsDataTable();
	dt.Print();
}
```

How do I add another column for `mykey` like so:

```
mykey, name, teamname, email, player
1234, asdf, b, c, 1
1234, asdf, b, c, 2
```

**Comment by gjvanvuuren** [comment]
In the following code, for example, the columns aren't renamed, and replacing `.WithJSONPath` with `.WithField` doesn't work either:

```
var conf = new ChoJSONRecordConfiguration
{
	NestedKeySeparator = null,
	UseNestedKeyFormat = true,
	FlattenNode = true
};

using (var r = ChoJSONReader.LoadText(json, conf)
	.WithField("mykey", jsonPath: "$.mykey", fieldName: "mykey")
	.WithField("username", jsonPath: "$.user.name", fieldName: "username")
	.WithField("player", jsonPath: "$.user.players[*]", fieldName: "player")
)
{
	var dt = r.AsDataTable();
	dt.Print();
}
```

Which results in:

```
mykey_0,username_0,player_0,player_1
1234,asdf,1,2
```

**Comment by gjvanvuuren** [comment]
Thanks again for the support.

My scenario is the following: Whereas the previous suggestions can be applied in individual cases, I have to process (tabularize) about 20 different json structures sent by machines. I do this by declaring metadata (source name, target name, data type, ...) in a configuration file for each of the messages, and then using code generation and ChoETL to do the plumbing.

This means I have to find a consistent way in which to process all different messages. 

- Your first approach means that I'll have to define the flattening using linq for each individual message, which will get very messy. 
- The second approach works well for the simpler messages, but not the complicated ones. And it's not compatible with the `.WithField` interface. I use the latter do define the data types and the target names.

So far I've had the best results over **all** messages by using `FlattenNode = true` in combination with `.WithField` like so:

```
var json = @"{
  ""Message"": ""MessageX"",
  ""TimestampLocal"": ""2022-02-02T02:02:34.2830859+01:00"",
  ""TimestampGlobal"": ""2022-02-02T00:58:58.8800398Z"",
  ""PickerId"": ""1"",
  ""DestinationComponentId"": 0,
  ""SourceComponentId"": 0,
  ""ActionList"": {
    ""ActionId"": [
      4428,
	  4429
    ]
  },
  ""TransactionId"": ""999bd293-b040-4e88-a89f-180cf9307597"",
  ""TransactionType"": ""none""
}";

var conf = new ChoJSONRecordConfiguration
{
	Culture = System.Globalization.CultureInfo.InvariantCulture,
	FlattenNode = true,
	ErrorMode = ChoErrorMode.IgnoreAndContinue,
};

using
(
	var r = ChoJSONReader.LoadText(json, conf)
		.WithField("TimestampGlobal", jsonPath: "$.TimestampGlobal", fieldType: typeof(DateTime), fieldName: "TimestampGlobal")
		.WithField("TimestampLocal", jsonPath: "$.TimestampLocal", fieldType: typeof(DateTime), fieldName: "TimestampLocal")
		.WithField("PickerId", jsonPath: "$.PickerId", fieldType: typeof(string), fieldName: "PickerId")
		.WithField("DestinationComponentId", jsonPath: "$.DestinationComponentId", fieldType: typeof(int), fieldName: "DestinationComponentId")
		.WithField("SourceComponentId", jsonPath: "$.SourceComponentId", fieldType: typeof(int), fieldName: "SourceComponentId")
		.WithField("ActionListIds", valueConverter: e => e != null ? String.Join(",", (IEnumerable<object>)e) : null, jsonPath: "$.ActionList.ActionId[*]", fieldType: typeof(string), fieldName: "ActionListIds")
		.WithField("TransactionId", jsonPath: "$.TransactionId", fieldType: typeof(Guid), fieldName: "TransactionId")
		.WithField("TransactionType", jsonPath: "$.TransactionType", fieldType: typeof(string), fieldName: "TransactionType")
)
{
	var dt = r.AsDataTable();
	Console.WriteLine(dt);
}
```

But then I'm stuck with having to use `valueConverter:` to squeeze `ActionId` into a 1x1 instead of a nx1. Would it be possible to use the `valueConverter` argument to embed some linq transformation, like the one in your first suggestion?

**Comment by Cinchoo** [comment]
well, the first approach `FlattenNode` only works one level expansion of the each node. 

With your requirement, you should go for `WithField` along with `Flatten`/`FlattenBy` operation to build the datatable.

```
          var json = @"{
""Message"": ""MessageX"",
""TimestampLocal"": ""2022-02-02T02:02:34.2830859+01:00"",
""TimestampGlobal"": ""2022-02-02T00:58:58.8800398Z"",
""PickerId"": ""1"",
""DestinationComponentId"": 0,
""SourceComponentId"": 0,
""ActionList"": {
  ""ActionId"": [
    4428,
	4429
  ]
},
""TransactionId"": ""999bd293-b040-4e88-a89f-180cf9307597"",
""TransactionType"": ""none""
}";

          var conf = new ChoJSONRecordConfiguration
          {
              Culture = System.Globalization.CultureInfo.InvariantCulture,
              //FlattenNode = true,
              ErrorMode = ChoErrorMode.IgnoreAndContinue,
          };

          using
          (
              var r = ChoJSONReader.LoadText(json, conf)
                  .WithField("TimestampGlobal", jsonPath: "$.TimestampGlobal", fieldType: typeof(DateTime), fieldName: "TimestampGlobal")
                  .WithField("TimestampLocal", jsonPath: "$.TimestampLocal", fieldType: typeof(DateTime), fieldName: "TimestampLocal")
                  .WithField("PickerId", jsonPath: "$.PickerId", fieldType: typeof(string), fieldName: "PickerId")
                  .WithField("DestinationComponentId", jsonPath: "$.DestinationComponentId", fieldType: typeof(int), fieldName: "DestinationComponentId")
                  .WithField("SourceComponentId", jsonPath: "$.SourceComponentId", fieldType: typeof(int), fieldName: "SourceComponentId")
                  //.WithField("ActionListIds", valueConverter: e => e != null ? String.Join(",", (IEnumerable<object>)e) : null, jsonPath: "$.ActionList.ActionId[*]", fieldType: typeof(string), fieldName: "ActionListIds")
                  .WithField("ActionListIds", jsonPath: "$.ActionList.ActionId[*]", fieldName: "ActionListIds")
                  .WithField("TransactionId", jsonPath: "$.TransactionId", fieldType: typeof(Guid), fieldName: "TransactionId")
                  .WithField("TransactionType", jsonPath: "$.TransactionType", fieldType: typeof(string), fieldName: "TransactionType")
          )
          {
              var dt = r.FlattenBy().AsDataTable();
              dt.Print();
          }
```
Sample fiddle: https://dotnetfiddle.net/AmSy3G

Hope it helps.

**Comment by gjvanvuuren** [comment]
Your proposal for using `FlattenBy` works fine for the one message type 'A', but testing it with messages containing nested nodes (message type 'B'), I have to switch to using the `FlattenNode = true` configuration setting. As you suggested, in both cases I can use `WithField`. I've coded logic to switch between either type A or type B. So far so good.

But then I ran into an unexpected issue. The machines don't always send all the fields. Normally ChoETL handels this elegantly. But to use the above example, if the `ActionList` node is missing, calling `FlattenBy()` returns an empty result. This is probably because the json then consists only of a single node and is already flattened.

So my question is: How can I do a runtime check to determine if a json can be flattened. Or is this something that ChoETL can or could handle internally?
